ql/src/java/org/apache/hadoop/hive/ql/parse/repl/metric/MetricCollector.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql.parse.repl.metric;
+
+import org.apache.hadoop.hive.ql.parse.SemanticException;
+import org.apache.hadoop.hive.ql.parse.repl.metric.event.ReplicationMetric;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * MetricCollector.
+ * In memory collection of metrics
+ */
+public class MetricCollector {
+  private Logger LOG = LoggerFactory.getLogger(MetricCollector.class);
+  private Map<Long, ReplicationMetric> metricMap = new ConcurrentHashMap<>();
+  private final long maxSize;
+
+  private static volatile MetricCollector instance;
+
+  private MetricCollector(long maxSize){
+    this.maxSize = maxSize;
+  }
+
+  public static MetricCollector getInstance(long maxSize) {
+    if (instance == null) {
+      synchronized (MetricCollector.class) {
+        if (instance == null) {
+          instance = new MetricCollector(maxSize);
+        }
+      }
+    }
+    return instance;
+  }
+
+  public void addMetric(ReplicationMetric replicationMetric) throws SemanticException {
```

Comment:
How many concurrent thread at a time can call this add metric? Is data copy progress part of this metric?

Reply:
Yes that is why metric map is a concurrent hash map

Reply:
Initially it was just a put so concurrent hashmap was sufficient. Now it has the size check etc so needs to be sync

Reply:
yes, also, we are doing draining out the hashmap data to another data structure, there is a race condition at that point too.